eval/src/expr/binary_op.rs

```diff
@@ -0,0 +1,289 @@
+            if rhs.is_zero() {
+                Err(RuntimeErrorKind::DivisionByZero)?
+            }
+            match (&lhs, &rhs) {
+                (JmlValue::Float(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float(lhs.0 % rhs.0)),
+                (JmlValue::Float(lhs), JmlValue::Int(rhs)) => Ok(JmlValue::float(lhs.0 % rhs.0 as f64)),
+                (JmlValue::Int(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float(lhs.0 as f64 % rhs.0 )),
+                (JmlValue::Int(lhs), JmlValue::Int(rhs)) =>Ok(JmlValue::int(lhs.0 % rhs.0)),
+                _ => Err(TypeErrorKind::InvalidBinaryOperator {
+                    operator: "%".to_owned(),
+                    left: lhs.type_of(),
+                    right: rhs.type_of(),
+                })?,
+            }
+
+        }
+    };
+
+    ($func_name:ident, ^) => {
+        fn $func_name<'a>(lhs: JmlValue, rhs: JmlValue) -> anyhow::Result<JmlValue<'a>> {
+            match (&lhs, &rhs) {
+                (JmlValue::Float(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float(lhs.0.powf(rhs.0))),
+                (JmlValue::Float(lhs), JmlValue::Int(rhs)) => Ok(JmlValue::float(lhs.0.powf(rhs.0 as f64))),
+                (JmlValue::Int(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float((lhs.0 as f64).powf(rhs.0))),
+                (JmlValue::Int(lhs), JmlValue::Int(rhs)) => {
+                    match lhs.0.checked_pow(rhs.0 as u32) {
+                        Some(result) => Ok(JmlValue::int(result)),
+                        None => Err(RuntimeErrorKind::Overflow)?,
+                    }
+                }
+                _ => Err(TypeErrorKind::InvalidBinaryOperator {
+                    operator: "^".to_string(),
+                    left: lhs.type_of(),
+                    right: rhs.type_of(),
+                }.into()),
+            }
+        }
+    };
+    ($func_name:ident, $operator:tt, $op_str:expr) => {
+        fn $func_name<'a>(lhs: JmlValue, rhs: JmlValue) -> Result<JmlValue<'a>, TypeErrorKind> {
+            match (&lhs, &rhs) {
+                (JmlValue::Float(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float(lhs.0 $operator rhs.0)),
+                (JmlValue::Float(lhs), JmlValue::Int(rhs)) => Ok(JmlValue::float(lhs.0 $operator rhs.0 as f64)),
+                (JmlValue::Int(lhs), JmlValue::Float(rhs)) => Ok(JmlValue::float(lhs.0 as f64 $operator rhs.0 )),
+                (JmlValue::Int(lhs), JmlValue::Int(rhs)) =>Ok(JmlValue::int(lhs.0 $operator rhs.0)),
+                _ => Err(TypeErrorKind::InvalidBinaryOperator {
+                    operator: $op_str.to_string(),
+                    left: lhs.type_of(),
+                    right: rhs.type_of(),
+                }),
+            }
+        }
+    };
+}
+
+arithmetic_op!(add, +, "+");
+arithmetic_op!(subtract, -, "-");
+arithmetic_op!(multiply, *, "*");
+arithmetic_op!(mod_op, %);
+arithmetic_op!(divide, /);
+arithmetic_op!(pow, ^);
+
+fn concat<'source>(
```

Comment:
Isn't this part of stdlib?

Reply:
kind of ... but there is `++` op that has same functionality